README
Text::DHCPLeases Perl Module


INSTALLATION

You have two options:

perl Build.PL
./Build
./Build test
./Build install

or (if you have GNU make):
perl Makefile.PL
make
make test
make install


DOCUMENTATION

Documentation in POD format included with code.  Try the following:

	 perldoc lib/Text/DHCPLeases.pm
	 perldoc lib/Text/DHCPLeases/Object.pm
	 perldoc lib/Text/DHCPLeases/Object/Iterator.pm

COPYRIGHT AND LICENCE

Copyright (c) 2007-2010, Carlos Vicente <cvicente@cpan.org>. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for <DISTRO>

0.9  2009/08/10 
        - Added License text in all files
        - Removed requirement for version.pm

0.8  2008/12/07
        - Added support for ddns-rev-name, ddns-txt and ddns-fwd-name statements, based on patch
          by Franck Joncourt.

0.7  2008/11/06
	- Changed regex to deal with 'ends never' statement
 	- Warn instead of die when statement not recognized

0.6  2008/07/03
	- Deal with incomplete hardware address line

0.5  2008/06/05
	- Fixed inconsistent licensing

0.4  2008/03/26
	- Fixed bug in Iterator.pm which caused it to miss first object in list.

0.3  2008/03/13
	- Fixed regular expression in Text::DHCPLeases::Object::parse that failed when hardware-type value included non-word characters
          (i.e "token-ring").
        - Added version.pm to list of requisites in Build.PL

0.2  2007/12/20
	**** API CHANGED ****
	- Renamed Leases.pm to Object.pm.  This reflects changes in the code that allows it to deal with all possible kinds of declarations
	  in the leases file (was missing objects created by OMAPI).
	- Fixed bugs in parsing reported by John Lloyd.

0.1  2007/08/21
       Initial release.


Copyright (c) 2007-2010, Carlos Vicente <cvicente@cpan.org>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.


lib/Text/DHCPLeases.pm
package Text::DHCPLeases;

use warnings;
use strict;
use Carp;
use Text::DHCPLeases::Object;
use Text::DHCPLeases::Object::Iterator;
use vars qw($VERSION);
$VERSION = '0.9';

my $IPV4  = '\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}';

# Make sure to return 1
1;

=head1 NAME

Text::DHCPLeases - Parse DHCP leases file from ISC dhcpd.

=head1 SYNOPSIS

    use Text::DHCPLeases;

    my $leases = Text::DHCPLeases->new("/etc/dhcpd.leases");

    foreach my $obj ( $leases->get_objects ){
        print $obj->name;
        if ( $obj->binding_state eq 'active' ){
           ...
    }
    ...

=head1 DESCRIPTION

This module provides an object-oriented interface to ISC DHCPD leases files.  
The goal is to objectify all declarations, as defined by the ISC dhcpd package man pages.

This interface is useful for analyzing, reporting, converting lease files, 
or as a tool for other applications that need to import dhcpd lease data structures.

=head1 CLASS METHODS
=cut

############################################################
=head2 new - Class Constructor

  Arguments:
    Hash with the following keys:
    file  -  Leases file path
  Returns:
    Text::DHCPLeases object
  Examples:
    Text::DHCPLeases->new(file=>"/etc/dhcpd.leases");
=cut
sub new{
    my ($proto, %argv) = @_;
    croak "Missing required parameters: file" unless defined $argv{file};
    my $class = ref($proto) || $proto;
    my $self = {};
    bless $self, $class;
    $self->{_objects} = $self->_parse($argv{file});
    return $self;
}


=head1 INSTANCE METHODS
=cut


############################################################
=head2 get_objects - Get objects from leases file

  Arguments:
    Object attributes to match (optional)
  Returns:
    Array of Text::DHCPLeases::Lease objects, 
    or iterator depending on context.  
  Examples:
    my $it = $leases->get_objects(ip_address=>'192.168.0.1');
    while ( my $obj = $it->next ) ...
=cut
sub get_objects{
    my ($self, %argv) = @_;
    my @list;
    if ( %argv ){
	foreach my $obj ( @{$self->{_objects}} ){
	    my $match = 1;
	    foreach my $key ( keys %argv ){
		if ( !defined $obj->$key || $obj->$key ne $argv{$key} ){
		    $match = 0;
		    last;
		}
	    }
	    push @list, $obj if $match;
	}
    }else{
	# Use 'all' array to get real order from file
	@list = @{$self->{_objects}};
    }
    wantarray? @list : DHCPLeases::Object::Iterator->new(\@list);
}

############################################################
=head2 print - Print all lease objects contents as a formatted string

  Arguments:
    None
  Returns:
    Formatted String
  Examples:
    print $leases->print;
=cut
sub print{
    my ($self) = @_;
    my $out = "";
    foreach my $obj ( $self->get_objects ){
	$out .= $obj->print;
    }
    return $out;
}

############################################################
#
# ********* PRIVATE METHODS **********
#
############################################################


############################################################
# _parse - Populate array of objects after reading file
#
# Arguments:
#    filename
# Returns:
#    Hash reference.  
#    Key:   declaration header
#    Value: reference to array with all objects
#
sub _parse {
    my ($self, $file) = @_;
    my @objects;
    my $declist = $self->_get_decl($file);
    foreach my $decl ( @$declist ){
	my $header = $decl->{header};
	my $lines  = $decl->{lines};
	my $obj;
	if ( $header =~ /^(lease|host|group|subgroup|failover peer)/ ){
	    my $obj_data = Text::DHCPLeases::Object->parse($lines);
	    $obj = Text::DHCPLeases::Object->new(%$obj_data);
	    push @objects, $obj;	
	}else{
	    croak "Text::DHCPLeases::_parse Error: Declaration header not recognized: $header\n";
	}
    }
    return \@objects;
}

############################################################
# _get_decl - Parse file and return all declarations
#
# Arguments:
#    filename
# Returns:
#    Array ref of hashrefs.  
#    
sub _get_decl {
    my ($self, $file) = @_;
    open(FILE, "<$file") or croak "Can't open file $file: $!\n";
    my @list;
    my $lines = [];
    my $header;
    my $open = 0;
    my $decl;
    while ( <FILE> ){
	my $line = $_;
	$line =~ s/^\s*(.*)\s*$/$1/;
	next if ( $line =~ /^#|^$/ );
	if ( !$open && $line =~ /^(.*) \{$/ ){
	    $decl = {};
	    $header = $1;
	    $decl->{header} = $header;
  	    $open   = 1;
	    $lines  = [];
	    push @$lines, $line;
	    next;
	}
	if ( $open ){
	    if ( $line =~ /^\}$/ ){
		$open = 0;
		$decl->{lines}  = $lines;
		push @list, $decl;
		$header = "";
		push @$lines, $line;
	    }else{
		push @$lines, $line;
	    }
	}
    }
    close(FILE);
    return \@list;
}


=head1 BUGS AND LIMITATIONS

Correct parsing of leases files depends on changes made to the format of
said files by the authors of the ISC DHCPD package.  This module was tested
against leases files generated by ISC DHCPD version 3.1.0.  In addition, I
do not have access to leases file with all possible declarations and statements,
so parsing could be broken in some circumstances.  Patches are welcome.

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-Text-DHCPleases@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Carlos Vicente  <cvicente@cpan.org>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2007-2010, Carlos Vicente <cvicente@cpan.org>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
=cut


lib/Text/DHCPLeases/FPS.pm
package Text::DHCPLeases::FPS;

use warnings;
use strict;
use Carp;
use Class::Struct;

use version; our $VERSION = qv('0.1');

my $IPV4  = '\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}';

# weekday year/month/day hour:minute:second
my $DATE  = '\d+ \d{4}\/\d{2}\/\d{2} \d{2}:\d{2}:\d{2}';

=head1 NAME

Text::DHCPLeases::FPS - Failover Peer State class

=head1 SYNOPSIS

my $fps = Text::DHCPLeases::FPS->new(%data);
print $fps->my_state->{state};
print $fps->partner_state->{state};

=head1 DESCRIPTION

Failover Peer State objects and their operations

=cut

struct (
'name'                 => '$',
'my_state'             => '%',
'partner_state'        => '%',
'mclt'                 => '$',
);

=head1 CLASS METHODS

=head2 new - Constructor
  Arguments:
    name
    my_state
    partner_state
    mclt
  Returns:
    New Text::DHCPLeases::FPS object
  Examples:
    my $fps = Text::DHCPLeases::FPS->new(%data);
    
=cut
############################################################
=head2 parse - Parse failover peer state declaration

  Arguments:
    Array ref with declaration lines
  Returns:
    Hash reference.  
  Examples:
    
my $text = 'failover peer "dhcp-peer" state {
 my state communications-interrupted at 2 2007/08/14 21:10:00;
 partner state normal at 2 2007/08/14 20:51:22;
 mclt 3600;
}';

my @lines = split /\n/, $text;
my $fps_data = Text::DHCPLeases::Lease->parse(\@lines);
=cut
sub parse {
    my ($self, $lines) = @_;
    my %fps;
    for ( @$lines ){
	next if ( /^#|^$|\}$/ );
	if ( /failover peer \"(.*)\" state/ ){
	    $fps{name} = $1;
	}elsif ( /my state (.*) at ($DATE);/ ){
	    $fps{my_state}{state} = $1;
	    $fps{my_state}{date}  = $2;
	}elsif (/partner state (.*) at ($DATE);/ ){
	    $fps{partner_state}{state} = $1;
	    $fps{partner_state}{date}  = $2;
	}elsif (/mclt (\w+);/ ){
	    $fps{mclt} = $1;
	}else{
	    croak "Text::DHCPLeases::FPS::parse Error: Statement not recognized: $_\n";
	}
    }
    return \%fps;
}


=head1 INSTANCE METHODS
=cut

############################################################
=head2 print - Print formatted string with object contents

  Arguments:
    None
  Returns:
    Formatted String
  Examples:
    print $fps->print;
=cut
sub print{
    my ($self) = @_;
    my $out = "";
    $out .= sprintf("failover peer \"%s\" state {\n", $self->name);
    $out .= sprintf(" my state %s at %s;\n", $self->my_state->{state}, $self->my_state->{date});
    $out .= sprintf(" partner state %s at %s;\n", $self->partner_state->{state}, $self->partner_state->{date});
    $out .= sprintf(" mclt %s;\n", $self->mclt);
    $out .= "}\n";
    return $out;
}

# Make sure to return 1
1;

=head1 AUTHOR

Carlos Vicente  C<< <<cvicente@cpan.org>> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) <2007-2010>, Carlos Vicente C<< <<cvicente@cpan.org>> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
=cut


lib/Text/DHCPLeases/Object/Iterator.pm
package DHCPLeases::Object::Iterator;

use strict;
use warnings;
use vars qw($VERSION);
$VERSION = '0.9';

=head1 NAME

Text::DHCPLeases::Object::Iterator - Lease object iterator class

=head1 SYNOPSIS


=head1 DESCRIPTION
=cut


sub new {
    my ($proto, $list) = @_;
    my $class = ref($proto) || $proto;
    my $self = {};
    $self->{_list} = $list;
    $self->{_pos}  = 0;
    $self->{_size} = scalar @{$self->{_list}};
    bless $self, $class;
}

sub count { my ($self) = @_; return $self->{_size} };

sub first {
    my ($self) = @_;
    return $self->{_list}->[0];
}

sub last {
    my ($self) = @_;
    return $self->{_list}->[$self->{_size} - 1];
}

sub next {
    my ($self) = @_;
    return $self->{_list}->[$self->{_pos}++];
}

# Make sure to return 1
1;



=head1 AUTHOR

Carlos Vicente, <cvicente@cpan.org>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2007-2010, Carlos Vicente <cvicente@cpan.org>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
=cut



lib/Text/DHCPLeases/Lease.pm
package Text::DHCPLeases::Lease;

use warnings;
use strict;
use Carp;
use Class::Struct;

use version; our $VERSION = qv('0.1');

my $IPV4  = '\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}';

# weekday year/month/day hour:minute:second
my $DATE  = '\d+ \d{4}\/\d{2}\/\d{2} \d{2}:\d{2}:\d{2}';

=head1 NAME

Text::DHCPLeases::Lease - Lease class

=head1 SYNOPSIS

my $lease = Text::DHCPLeases::Lease->new(%lease_data);
print $lease->address;
print $lease->binding_state;

=head1 DESCRIPTION

Lease objects and their operations

=cut

struct (
'address'                 => '$',
'starts'                  => '$',
'ends'                    => '$',
'tstp'                    => '$',
'tsfp'                    => '$',
'atsfp'                   => '$',
'cltt'                    => '$',
'next_binding_state'      => '$',
'binding_state'           => '$',
'uid'                     => '$',
'client_hostname'         => '$',
'abandoned'               => '$',
'option_agent_circuit_id' => '$',
'option_agent_remote_id'  => '$',
'hardware'                => '%',
'set'                     => '%',
'on'                      => '%',
'bootp'                   => '$',
'resrved'                 => '$',
);

=head1 CLASS METHODS

=head2 new - Constructor

  Arguments:
    address
    starts
    ends
    tstp
    tsfp
    atsfp
    cltt
    next_binding_state
    binding_state
    uid
    client_hostname
    abandoned (flag)
    option_agent_circuit_id
    option_agent_remote_id
    hardware (hash)
    set (hash)
    on (hash)
    bootp (flag)
    reserved (flag)
  Returns:
    New Text::DHCPLeases::Lease object
  Examples:

    my $lease = Text::DHCPLeases::Leases->new(address => '192.168.1.10',
                                              starts  => '3 2007/08/15 11:34:58',
                                              ends    => '3 2007/08/15 11:44:58');
   
=cut

############################################################
=head2 parse - Parse lease declaration

Arguments:
   Array ref with declaration lines
Returns:
   Hash reference.  
  Examples:

    my $text = '
lease 192.168.254.55 {
  starts 3 2007/08/15 11:34:58;
  ends 3 2007/08/15 11:44:58;
  tstp 3 2007/08/15 11:49:58;
  tsfp 2 2007/08/14 21:24:19;
  cltt 3 2007/08/15 11:34:58;
  binding state active;
  next binding state expired;
  hardware ethernet 00:11:85:5d:4e:11;
  uid "\001\000\021\205]Nh";
  client-hostname "blah";
}';

my $lease_data = Text::DHCPLeases::Lease->parse($text);
=cut
sub parse{
    my ($self, $lines) = @_;
    my %lease;
    for ( @$lines ){
	next if ( /^#|^$|\}$/ );
	if ( /lease ($IPV4) / ){
	    $lease{address} = $1;
	}elsif ( /starts ($DATE);/ ){
	    $lease{starts} = $1;
	}elsif ( /ends ($DATE);/ ){
	    $lease{ends} = $1;
	}elsif ( /tstp ($DATE);/ ){
	    $lease{tstp} = $1;
	}elsif ( /tsfp ($DATE);/ ){
	    $lease{tsfp} = $1;
	}elsif ( /atsfp ($DATE);/ ){
	    $lease{atsfp} = $1;
	}elsif ( /cltt ($DATE);/ ){
	    $lease{cltt} = $1;
	}elsif ( /next binding state (\w+);/ ){
	    $lease{'next_binding_state'} = $1;
	}elsif ( /binding state (\w+);/ ){
	    $lease{'binding_state'} = $1;
	}elsif ( /uid (\".*\");/ ){
	    $lease{uid} = $1;
	}elsif ( /client-hostname (\".*\");/ ){
	    $lease{'client_hostname'} = $1;
	}elsif ( /abandoned;/ ){
	    $lease{abandoned} = 1;
	}elsif ( /hardware (\w+) (.*);/ ){
	    $lease{hardware}{'hardware-type'} = $1;
	    $lease{hardware}{'mac-address'}   = $2;
	}elsif ( /option agent.circuit-id (\".*\");/ ){
	    $lease{'option_agent_circuit_id'} = $1;
	}elsif ( /option agent.remote-id (\".*\");/ ){
	    $lease{'option_agent_remote_id'} = $1;
	}elsif ( /set (\w+) = (.*);/ ){
	    $lease{set}{$1} = $2;
	}elsif ( /on (.*) \{(.*)\};/ ){
	    my $events     = $1;
	    my @events = split /\|/, $events;
	    my $statements = $2;
	    my @statements = split /\n;/, $statements;
	    $lease{on}{events}     = @events;
	    $lease{on}{statements} = @statements;
	}elsif ( /bootp;/ ){
	    $lease{bootp} = 1;
	}elsif ( /reserved;/ ){
	    $lease{reserved} = 1;
	}else{
	    croak "Text::DHCPLeases::Lease::parse Error: Statement not recognized: $_\n";
	}
    }
    return \%lease;
}

=head1 INSTANCE METHODS
=cut

############################################################
=head2 print - Print formatted string with lease contents

  Arguments:
    None
  Returns:
    Formatted String
  Examples:
    print $lease->print;
=cut
sub print{
    my ($self) = @_;
    my $out = "";
    $out .= sprintf("lease %s {\n",   $self->address);
    $out .= sprintf("  starts %s;\n", $self->starts) if $self->starts;
    $out .= sprintf("  ends %s;\n",   $self->ends)   if $self->ends;
    $out .= sprintf("  tstp %s;\n",   $self->tstp)   if $self->tstp;
    $out .= sprintf("  tsfp %s;\n",   $self->tsfp)   if $self->tsfp;
    $out .= sprintf("  atsfp %s;\n",  $self->atsfp)  if $self->atsfp;
    $out .= sprintf("  cltt %s;\n",   $self->cltt)   if $self->cltt;
    $out .= sprintf("  binding state %s;\n",   $self->binding_state)   
	if $self->binding_state;
    $out .= sprintf("  next binding state %s;\n",   $self->next_binding_state)
	if $self->next_binding_state;
    $out .= sprintf("  hardware %s %s;\n", $self->hardware->{'hardware-type'}, 
		    $self->hardware->{'mac-address'}) if $self->hardware;
    $out .= sprintf("  uid %s;\n",    $self->uid)   if $self->uid;
    $out .= sprintf("  client-hostname %s;\n", $self->client_hostname) if $self->client_hostname;
    $out .= sprintf("  abandoned %s;\n", $self->abandoned) if $self->abandoned;
    $out .= sprintf("  option agent.circuit-id %s;\n", $self->option_agent_circuit_id) 
	if $self->option_agent_circuit_id;
    $out .= sprintf("  option agent.remote-id %s;\n", $self->option_agent_remote_id) 
	if $self->option_agent_remote_id;
    if ( defined $self->set ){
	foreach my $var ( keys %{ $self->set } ){
	    $out .= sprintf("  set %s = %s;\n", $var, $self->set->{$var});
	}
    }
    if ( $self->on && $self->on->{events} && $self->on->{statements} ){
	my $events = join '|', @{$self->on->{events}};
	my $statements = join '\n;', @{$self->on->{statements}};
	$out .= sprintf("  on %s { %s }", $events, $statements);

    }
    $out .= "}\n";
    return $out;
}

# Make sure to return 1
1;

=head1 AUTHOR

Carlos Vicente  C<< <<cvicente@cpan.org>> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) <2007-2010>, Carlos Vicente C<< <<cvicente@cpan.org>> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
=cut
